guides/create_an_app/2-channels.md

```diff
@@ -1,5 +1,161 @@
+
+```elixir
+defmodule WorkshopsAppWeb.GeneralChannel do
+  use WorkshopsAppWeb, :channel
+
+  def join("general:lobby", payload, socket) do
+    if authorized?(payload) do
+      {:ok, socket}
+    else
+      {:error, %{reason: "unauthorized"}}
+    end
+  end
+
+  # Channels can be used in a request/response fashion
+  # by sending replies to requests from the client
+  def handle_in("ping", payload, socket) do
+    {:reply, {:ok, payload}, socket}
+  end
+
+  # It is also common to receive messages from the client and
+  # broadcast to everyone in the current topic (general:lobby).
+  def handle_in("shout", payload, socket) do
+    broadcast socket, "shout", payload
+    {:noreply, socket}
+  end
+
+  # Add authorization logic here as required.
+  defp authorized?(_payload) do
+    true
+  end
+end
+```
+
+Now that we have generated some code let's look at what it does. First the `general_channel.ex` is the channel equivalent of a controller in Phoenix, being that it handles websocket requests instead of http requests. On line **2** `:channel` is referring to the Phoenix channel function from `user_socket.ex` file. It takes a parameter which is used as a namespace for the websocket connections. In this instance the name space is **"general:lobby"**.
+
+The next argument `WorkshopsAppWeb.GeneralChannel` is made up of two parts. The first part `WorkshopsAppWeb` is the name of the app with `-Web` as suffix indicating that we have separate directory containing all the abstractions related to communication with outside world . The second part is `GeneralChannel`. This is the name of the module that handles the logic related to any communication through this channel. In Phoenix, it’s conventional to namespace all your modules with your app name first as this prevents namespace errors.
+
+The file, `user_socket.ex`, is like the `router.ex` file, but for channels. The topic, for us it's `general`, allows clients, also known as web browsers, to subscribe to the channel. In the same way a path in a web request directs the request to the right controller, the topic directs the socket to the right module, in this case, `WorkshopsAppWeb.GeneralChannel`.
+
+Now, let’s look at the join function at line **4**. `def join ("general:lobby", payload, socket) do:` takes three arguments. The first argument, the name of the channel, **"general:lobby"**, makes it so that this function is only called when the client is joining the specific channel. The second argument, **payload** is the request from the user; it can contain auth credentials, a message or anything the user wants. Finally, it takes **socket**, which is the websocket connection. There’s call to `authorized?` function, which always returns true. Then, `{:ok, socket}` just returns a status and the websocket they are connecting to. **join** function always returns `{:ok, socket}` to allow all connections connect to the channel.
```

Comment:
Now, let’s look at the join function at line **4**. `def join ("general:lobby", payload, socket) do:` takes three arguments. The first argument, the name of the channel, **"general:lobby"**, makes it so that this function is only called when the client is joining the specific channel. The second argument, **payload** is the request from the user; it can contain auth credentials, a message or anything the user wants. Finally, it takes **socket**, which is the websocket connection. There’s **a** call to `authorized?` function, which always returns true. Then, `{:ok, socket}` just returns a status and the websocket they are connecting to. **join** function always returns `{:ok, socket}` to allow all connections **to** connect to the channel.